landlab/utils/distance_to_divide.py

```diff
@@ -0,0 +1,234 @@
+    ...     left_is_closed=True,
+    ...     right_is_closed=True,
+    ...     top_is_closed=True)
+    >>> fr = FlowAccumulator(mg, flow_director = 'MFD')
+    >>> fr.run_one_step()
+    >>> distance_to_divide = calculate_distance_to_divide(
+    ...     mg,
+    ...     add_to_grid=True,
+    ...     noclobber=False)
+    >>> mg.at_node['distance_to_divide']
+    array([ 0.,  3.,  3.,  0.,
+            0.,  2.,  2.,  0.,
+            0.,  1.,  1.,  0.,
+            0.,  0.,  0.,  0.,
+            0.,  0.,  0.,  0.])
+
+    The distance_to_divide utility can also work on irregular grids. For the
+    example we will use a Hexagonal Model Grid, a special type of Voroni Grid
+    that has regularly spaced hexagonal cells.
+
+    >>> from landlab import HexModelGrid, FIXED_VALUE_BOUNDARY, CLOSED_BOUNDARY
+    >>> from landlab.components import FlowAccumulator
+    >>> from landlab.utils.distance_to_divide import (
+    ...     calculate_distance_to_divide)
+    >>> dx = 1
+    >>> hmg = HexModelGrid(5,3, dx)
+    >>> _ = hmg.add_field('topographic__elevation',
+    ...                   hmg.node_x + np.round(hmg.node_y),
+    ...                   at = 'node')
+    >>> hmg.status_at_node[hmg.boundary_nodes] = CLOSED_BOUNDARY
+    >>> hmg.status_at_node[0] = FIXED_VALUE_BOUNDARY
+    >>> fr = FlowAccumulator(hmg, flow_director = 'D4')
+    >>> fr.run_one_step()
+    >>> distance_to_divide = calculate_distance_to_divide(
+    ...     hmg,
+    ...     add_to_grid=True,
+    ...     noclobber=False)
+    >>> hmg.at_node['distance_to_divide']
+    array([ 3.,  0.,  0.,
+         0.,  2.,  1.,  0.,
+       0.,  1.,  1.,  0.,  0.,
+         0.,   0.,  0.,  0.,
+            0.,  0.,  0.])
+    """
+    # check that flow__receiver nodes exists
+    if "flow__receiver_node" not in grid.at_node:
+        raise FieldError(
+            "A 'flow__receiver_node' field is required at the "
+            "nodes of the input grid."
+        )
+
+    if "flow__upstream_node_order" not in grid.at_node:
+        raise FieldError(
+            "A 'flow__upstream_node_order' field is required at the "
+            "nodes of the input grid."
+        )
+
+    if "drainage_area" not in grid.at_node:
+        raise FieldError(
+            "A 'flow__upstream_node_order' field is required at the "
+            "nodes of the input grid."
+        )
+
+    # get the reciever nodes, depending on if this is to-one, or to-multiple,
+    # we'll need to get a different at-node field.
+    if grid.at_node["flow__receiver_node"].size != grid.size("node"):
+        to_one = False
+    else:
+        to_one = True
+
+    flow__receiver_node = grid.at_node["flow__receiver_node"]
+    drainage_area = grid.at_node["drainage_area"]
+
+    # get the upstream node order
+    flow__upstream_node_order = grid.at_node["flow__upstream_node_order"]
+
+    # get downstream flow link lengths, result depends on type of grid.
+    if isinstance(grid, RasterModelGrid):
```

Comment:
Ultimately we need to expand the menu of rectilinear grid types to include cases with nonuniform dx, dy, and when that day comes, this will need to reference whatever base class applies...